src/components/requirements/RequirementDisplayComponent.tsx

```diff
@@ -0,0 +1,147 @@
+import { ChainIndicator } from "@/components/requirements/ChainIndicator";
+import { RequirementLink } from "@/components/requirements/RequirementLink";
+import { CHAINS, type SupportedChainID } from "@/config/chains";
+import { GearSix } from "@phosphor-icons/react/dist/ssr";
+import {
+  Requirement,
+  RequirementContent,
+  RequirementFooter,
+  RequirementImage,
+} from "./Requirement";
+
+import type { Rule } from "@/lib/schemas/rule";
+import { shortenHex } from "@/lib/shortenHex";
+import { Fragment } from "react";
+import { DataBlockWithCopy } from "./DataBlockWithCopy";
+
+const PLACEHOLDER_REGEX = /\{\{([^{}]+)\}\}/g;
+const ADDRESS_REGEX = /^0x[a-f0-9]{40}$/i;
+
+const convertTemplateText = (templateText: string, requirement: Rule) =>
+  templateText.replace(PLACEHOLDER_REGEX, (_match, rawKey) => {
+    const key = rawKey.trim() as keyof typeof requirement.config &
+      keyof typeof requirement.integration;
+
+    const usableKeyValues = {
+      ...requirement.config,
+      ...requirement.integration,
+    };
+
+    return key in usableKeyValues ? usableKeyValues[key] : key;
+  });
+
+const isSupportedChain = (chainId?: number): chainId is SupportedChainID =>
+  chainId ? !!CHAINS[chainId as SupportedChainID] : false;
+
+const RequirementNode = ({
+  node,
+  requirement,
+}: {
+  node: Rule["ui"]["nodes"][number];
+  requirement: Rule;
+}) => {
+  switch (node.type) {
+    case "TEXT":
+      return (
+        <span key={node.id}>
+          {convertTemplateText(node.value, requirement)}
+        </span>
+      );
+    case "CHAIN_INDICATOR":
+      // @ts-expect-error - TODO: create a map for chainId - chainName (uppercase) pairs
+      return isSupportedChain(node.value) ? (
+        <ChainIndicator key={node.id} chain={node.value} />
+      ) : null;
+    case "EXTERNAL_LINK":
+      return (
+        <RequirementLink
+          key={node.id}
+          href={convertTemplateText(node.href, requirement)}
+        >
+          {node.value}
+        </RequirementLink>
+      );
+    default:
+      return null;
```

Comment:
Throw meaningful error here for unsupported / unimplemented requirement type. Ideally I would have the FE implemented requirement types in an array and maybe construct an object out of it with `FunctionComponent` callbacks.

Reply:
These are just requirement UI nodes, not requirements 😉 In theory, we shouldn't run into the default case, but I added a fallback badge now.

For requirements, the worst case scenario is when we don't have a proper config in the `ui` prop, but in that case, we just fall back to the default requirement UI.